iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/StringLiteralDuplicatedCheck.java

```diff
@@ -0,0 +1,135 @@
+
+  public static final int THRESHOLD_DEFAULT = 3;
+  public static final int MINIMAL_LITERAL_LENGTH_DEFAULT = 5;
+
+  @RuleProperty(
+    key = "threshold",
+    defaultValue = "" + THRESHOLD_DEFAULT)
+  protected int threshold = THRESHOLD_DEFAULT;
+
+  @RuleProperty(
+    key = "minimal_literal_length",
+    defaultValue = "" + MINIMAL_LITERAL_LENGTH_DEFAULT)
+  protected int minimalLiteralLength = MINIMAL_LITERAL_LENGTH_DEFAULT;
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(File.class, this::visitFile);
+  }
+
+  private void visitFile(CheckContext ctx, File file) {
+    var stringVisitor = new StringVisitor();
+    file.statements().forEach(stringVisitor::scan);
+    stringVisitor.reportDuplicates(ctx);
+  }
+
+  private class StringVisitor extends TreeVisitor<TreeContext> {
+
+    private static final TreeContext DUMMY_TREE_CONTEXT = new TreeContext();
+    private final Map<String, List<StringLiteral>> sameLiteralOccurrences = new HashMap<>();
+
+    public StringVisitor() {
+      register(StringLiteral.class, (ctx, tree) -> visitLiteral(tree));
+    }
+
+    private void scan(Tree root) {
+      super.scan(DUMMY_TREE_CONTEXT, root);
+    }
+
+    private void visitLiteral(StringLiteral tree) {
+      if (!isIgnored(tree)) {
+        var value = tree.value();
+        if (!sameLiteralOccurrences.containsKey(value)) {
+          List<StringLiteral> occurrences = new ArrayList<>();
+          occurrences.add(tree);
+          sameLiteralOccurrences.put(value, occurrences);
+        } else {
+          sameLiteralOccurrences.get(value).add(tree);
+        }
+      }
+    }
+
+    private boolean isIgnored(StringLiteral stringLiteral) {
+      var value = stringLiteral.value();
+      return value.length() < minimalLiteralLength
+        || ALLOWED_DUPLICATED_LITERALS.matcher(value).find()
+        || isResourceTypeAndApiVersionField(stringLiteral)
+        || isResourceId(stringLiteral);
+    }
+
+    private static boolean isResourceTypeAndApiVersionField(StringLiteral stringLiteral) {
+      Tree parent = stringLiteral.parent();
+      if (parent instanceof ResourceDeclaration resource) {
+        return stringLiteral.value().contains("@") || resource.type() == stringLiteral || resource.version() == stringLiteral;
+      }
+      return false;
+    }
+
+    private static boolean isResourceId(StringLiteral stringLiteral) {
+      Tree parent = stringLiteral.parent();
+      return parent instanceof FunctionCall functionCall && "resourceId".equals(functionCall.name().value());
+    }
+
+    private void reportDuplicates(CheckContext ctx) {
+      for (Map.Entry<String, List<StringLiteral>> literalOccurrences : sameLiteralOccurrences.entrySet()) {
+        List<StringLiteral> occurrences = literalOccurrences.getValue();
+        if (occurrences.size() >= threshold) {
+          var literal = literalOccurrences.getKey();
+          var message = String.format(MESSAGE, literal, occurrences.size());
```

Comment:
Possible (but not necessary) to use new syntax:  `var message = MESSAGE.formatted(literal, occurrences.size());`

Reply:
I wonder, why we don't have a Java rule for this.